### DISCUSSION TITLE
zinit load 'zoftmoth/zsh-vim-mode' is not allowing "^R" to be rebound to another widget

### DISCUSSION THREAD

**Discussion by rasivasu** [zdharma-continuum/zinit#596] [Q&A]
When I had the following code at the top of .zshrc

```zsh
zinit ice wait lucid atload'MODE_CURSOR_VIINS="bar"; vim-mode-cursor-init-hook' 
zinit load softmoth/zsh-vim-mode
```

and at the bottom,

```zsh
eval "$(mcfly init zsh)"

'softmoth/zsh-vim-mode' kept rebinding "^R" to a different widget.  How do I instruct this plugin to not attempt to rebind "^R" after loading?

When I added `set -x` at the top of .zshrc, I see that the zinit code was continuously running every few milliseconds.

**Comment by vladdoster** [comment]
@rasivasu,

What is `mcfly`?

What are you trying to bind `^R` to?

**Reply by pschmitt** [reply]
I believe it's this thing: https://github.com/cantino/mcfly

**Comment by vladdoster** [comment]
> I believe it's this thing: https://github.com/cantino/mcfly

lol, we also [test it](https://github.com/zdharma-continuum/zinit/blob/0ba778ac734e33c960fe08bbd56a351b1b86dcd4/tests/gh-r.zunit#L550-L554) 

```
@test 'mcfly' { # Fly through your shell history. Great Scott
  run zinit for @cantino/mcfly; assert $state equals 0
  local mcfly="$ZBIN/mcfly"; assert "$mcfly" is_executable
  run $mcfly --version; assert $state equals 0
}
```

**Comment by pschmitt** [comment]
You might want to take a look at the `bindmap` ice (see the README).

Or you could just unbind the key explicitly in the `atload` ice for your plugin.

**Comment by vladdoster** [comment]
@rasivasu,

Try the following recipe:

```zsh
zinit light @zdharma-continuum/zinit-annex-bin-gem-node;
zinit for \
  @softmoth/zsh-vim-mode \
       atinit'eval "$(mcfly init zsh)"' \
       atload'mcfly_prompt_command' \
       from'gh-r' \
       sbin'mcfly' \
  @cantino/mcfly;
```

Demonstrating `^R` is bound to mcfly and vim mode still works. It works, but video is too large to share. 

<img width="1729" alt="  " src="https://github.com/zdharma-continuum/zinit/assets/10052309/fe41709f-efc4-4c1f-8370-1bc611715f1f">

**Reply by rasivasu** [reply]
@vladdoster Earlier I didn't have Cargo installed and installed it for adding the above configuration to .zshrc.  Now I am getting the following error.  Can you please point me in the right direction?

![image](https://github.com/zdharma-continuum/zinit/assets/1153582/faeaa8c4-e173-456a-9e37-450808f1b661)

**Reply by vladdoster** [reply]
@rasivasu,

Please share the configuration. There is a syntax error somewhere.

**Reply by rasivasu** [reply]
@vladdoster Following is the configuration in which I inserted the above given recipe.  I source this file from .zshrc.  I pasted your recipe above the line containing 'marlonrichert/zsh-autocomplete'.

```zsh
### Added by Zinit's installer
if [[ ! -f $HOME/.local/share/zinit/zinit.git/zinit.zsh ]]; then
    print -P "%F{33} %F{220}Installing %F{33}ZDHARMA-CONTINUUM%F{220} Initiative Plugin Manager (%F{33}zdharma-continuum/zinit%F{220})…%f"
    command mkdir -p "$HOME/.local/share/zinit" && command chmod g-rwX "$HOME/.local/share/zinit"
    command git clone https://github.com/zdharma-continuum/zinit "$HOME/.local/share/zinit/zinit.git" && \
        print -P "%F{33} %F{34}Installation successful.%f%b" || \
        print -P "%F{160} The clone has failed.%f%b"
fi

source "$HOME/.local/share/zinit/zinit.git/zinit.zsh"

# Load a few important annexes, without Turbo
zinit wait lucid for \
    zdharma-continuum/zinit-annex-as-monitor \
    zdharma-continuum/zinit-annex-bin-gem-node \
    zdharma-continuum/zinit-annex-patch-dl \
    zdharma-continuum/zinit-annex-rust

### End of Zinit's installer chunk

# 'zdharma-continuum/fast-syntax-highlighting' Feature rich syntax highlighting for zsh
zinit wait lucid for \
 atinit"ZINIT[COMPINIT_OPTS]=-C; zicompinit; zicdreplay" \
    zdharma-continuum/fast-syntax-highlighting \
 atload"_zsh_autosuggest_start" \
 atinit"bindkey '^_' autosuggest-execute;bindkey '^ ' autosuggest-accept;" \
    @zsh-users/zsh-autosuggestions \
 blockf atpull'zinit creinstall -q .' \
    @zsh-users/zsh-completions

# 'marlonrichert/zsh-autocomplete' - Real-time type-ahead completion for Zsh. Asynchronous find-as-you-type autocompletion.
# zinit load marlonrichert/zsh-autocomplete

# 'agkozak/zsh-z' - jump quickly to frequently and recently accessed folders
zinit load agkozak/zsh-z

# Refer the documentation at https://github.com/softmoth/zsh-vim-mode
# 'softmoth/zsh-vim-mode' - Friendly bindings for ZSH's vi mode
# zinit ice wait lucid atload'MODE_CURSOR_VIINS="bar"; vim-mode-cursor-init-hook' 
# zinit load softmoth/zsh-vim-mode

# softmoth/zsh-vim-mode has been commented since zsh_init module is loaded repeatedly via a scheduler
# and this is rebinding "^R" to 

MODE_CURSOR_VIINS="#00ff00 blinking bar"
MODE_CURSOR_REPLACE="$MODE_CURSOR_VIINS #ff0000"
MODE_CURSOR_VICMD="green block"
MODE_CURSOR_SEARCH="#ff00ff steady underline"
MODE_CURSOR_VISUAL="$MODE_CURSOR_VICMD steady bar"
MODE_CURSOR_VLINE="$MODE_CURSOR_VISUAL #00ffff"

# Docker
zinit ice as"completion"
zinit snippet https://github.com/docker/cli/blob/master/contrib/completion/zsh/_docker

zinit ice from"gh-r" as"program" mv"docker* -> docker-compose"
zinit light docker/compose

```

**Comment by vladdoster** [comment]
@rasivasu,


> @vladdoster Following is the configuration in which I inserted the above given recipe. I source this file from .zshrc. I pasted your recipe above the line containing 'marlonrichert/zsh-autocomplete'.
> 
> ```shell
> ### Added by Zinit's installer
> if [[ ! -f $HOME/.local/share/zinit/zinit.git/zinit.zsh ]]; then
>     print -P "%F{33} %F{220}Installing %F{33}ZDHARMA-CONTINUUM%F{220} Initiative Plugin Manager (%F{33}zdharma-continuum/zinit%F{220})…%f"
>     command mkdir -p "$HOME/.local/share/zinit" && command chmod g-rwX "$HOME/.local/share/zinit"
>     command git clone https://github.com/zdharma-continuum/zinit "$HOME/.local/share/zinit/zinit.git" && \
>         print -P "%F{33} %F{34}Installation successful.%f%b" || \
>         print -P "%F{160} The clone has failed.%f%b"
> fi
> 
> source "$HOME/.local/share/zinit/zinit.git/zinit.zsh"
> 
> # Load a few important annexes, without Turbo
> zinit wait lucid for \
>     zdharma-continuum/zinit-annex-as-monitor \
>     zdharma-continuum/zinit-annex-bin-gem-node \
>     zdharma-continuum/zinit-annex-patch-dl \
>     zdharma-continuum/zinit-annex-rust
> 
> ### End of Zinit's installer chunk
> 
> # 'zdharma-continuum/fast-syntax-highlighting' Feature rich syntax highlighting for zsh
> zinit wait lucid for \
>  atinit"ZINIT[COMPINIT_OPTS]=-C; zicompinit; zicdreplay" \
>     zdharma-continuum/fast-syntax-highlighting \
>  atload"_zsh_autosuggest_start" \
>  atinit"bindkey '^_' autosuggest-execute;bindkey '^ ' autosuggest-accept;" \
>     @zsh-users/zsh-autosuggestions \
>  blockf atpull'zinit creinstall -q .' \
>     @zsh-users/zsh-completions
> 
> # 'marlonrichert/zsh-autocomplete' - Real-time type-ahead completion for Zsh. Asynchronous find-as-you-type autocompletion.
> # zinit load marlonrichert/zsh-autocomplete
> 
> # 'agkozak/zsh-z' - jump quickly to frequently and recently accessed folders
> zinit load agkozak/zsh-z
> 
> # Refer the documentation at https://github.com/softmoth/zsh-vim-mode
> # 'softmoth/zsh-vim-mode' - Friendly bindings for ZSH's vi mode
> # zinit ice wait lucid atload'MODE_CURSOR_VIINS="bar"; vim-mode-cursor-init-hook' 
> # zinit load softmoth/zsh-vim-mode
> 
> # softmoth/zsh-vim-mode has been commented since zsh_init module is loaded repeatedly via a scheduler
> # and this is rebinding "^R" to 
> 
> MODE_CURSOR_VIINS="#00ff00 blinking bar"
> MODE_CURSOR_REPLACE="$MODE_CURSOR_VIINS #ff0000"
> MODE_CURSOR_VICMD="green block"
> MODE_CURSOR_SEARCH="#ff00ff steady underline"
> MODE_CURSOR_VISUAL="$MODE_CURSOR_VICMD steady bar"
> MODE_CURSOR_VLINE="$MODE_CURSOR_VISUAL #00ffff"
> 
> # Docker
> zinit ice as"completion"
> zinit snippet https://github.com/docker/cli/blob/master/contrib/completion/zsh/_docker
> 
> zinit ice from"gh-r" as"program" mv"docker* -> docker-compose"
> zinit light docker/compose
> ```

This configuration does not have the snippet I suggested...? Did my snippet work?

```zsh
zinit light @zdharma-continuum/zinit-annex-bin-gem-node;
zinit for \
  @softmoth/zsh-vim-mode \
       atinit'eval "$(mcfly init zsh)"' \
       atload'mcfly_prompt_command' \
       from'gh-r' \
       sbin'mcfly' \
  @cantino/mcfly;
```

If so, please consider closing this issue and open a new one or start a discussion for other issues.

Additionally, to restart zsh & zinit cleanly, run the command:

```zsh
exec zsh
```

---

Try the following config:

Use this config exactly, do not modify it.

```zsh
if [[ ! -f $HOME/.local/share/zinit/zinit.git/zinit.zsh ]]; then
    print -P "%F{33} %F{220}Installing %F{33}ZDHARMA-CONTINUUM%F{220} Initiative Plugin Manager (%F{33}zdharma-continuum/zinit%F{220})…%f"
    if command mkdir -p "$HOME/.local/share/zinit"; then
        command chmod g-rwX "$HOME/.local/share/zinit"
    fi
    if command git clone https://github.com/zdharma-continuum/zinit "$HOME/.local/share/zinit/zinit.git"; then
        print -P "%F{33} %F{34}Installation successful.%f%b"
    else
        print -P "%F{160} The clone has failed.%f%b"
    fi
fi

source "$HOME/.local/share/zinit/zinit.git/zinit.zsh"
autoload _zinit && (( ${+_comps} )) && _comps[zinit]=_zinit

zinit for @zdharma-continuum/zinit-annex-{'as-monitor','binary-symlink','patch-dl','rust'}

MODE_CURSOR_VIINS="#00ff00 blinking bar"
MODE_CURSOR_REPLACE="$MODE_CURSOR_VIINS #ff0000"
MODE_CURSOR_VICMD="green block"
MODE_CURSOR_SEARCH="#ff00ff steady underline"
MODE_CURSOR_VISUAL="$MODE_CURSOR_VICMD steady bar"
MODE_CURSOR_VLINE="$MODE_CURSOR_VISUAL #00ffff"

zinit light-mode for \
      atinit'zicompinit;' \
    @marlonrichert/zsh-autocomplete \
    @softmoth/zsh-vim-mode

zinit ice as"completion"
zinit snippet https://github.com/docker/cli/blob/master/contrib/completion/zsh/_docker

zinit light-mode from"gh-r" for \
      lbin'!*->docker-compose' \
    @docker/compose \
      atinit'eval "$(mcfly init zsh)"' \
      atload'mcfly_prompt_command' \
      lbin'!mcfly' \
    @cantino/mcfly

zinit wait'!' lucid for \
      atinit"zicdreplay" \
    @zdharma-continuum/fast-syntax-highlighting \
      atload"_zsh_autosuggest_start" \
      atinit"bindkey '^_' autosuggest-execute;bindkey '^ ' autosuggest-accept;" \
    @zsh-users/zsh-autosuggestions \
      atpull'zinit creinstall -q .' \
      blockf \
    @zsh-users/zsh-completions
```

For testing, remove current install:

```zsh
rm -rf "${HOME}"/.local/share/zinit
exec zsh
```

**Comment by rasivasu** [comment]
@vladdoster <https://github.com/vladdoster>

I backed up my file and created a new file with your configuration.  I am
getting the following error.

[image: image.png]

Regards,
Ravi

On Sat, Dec 16, 2023 at 5:40 PM vladislav doster ***@***.***>
wrote:

> @rasivasu <https://github.com/rasivasu>,
>
> @vladdoster <https://github.com/vladdoster> Following is the
> configuration in which I inserted the above given recipe. I source this
> file from .zshrc. I pasted your recipe above the line containing
> 'marlonrichert/zsh-autocomplete'.
>
> ### Added by Zinit's installerif [[ ! -f $HOME/.local/share/zinit/zinit.git/zinit.zsh ]]; then
>     print -P "%F{33} %F{220}Installing %F{33}ZDHARMA-CONTINUUM%F{220} Initiative Plugin Manager (%F{33}zdharma-continuum/zinit%F{220})…%f"
>     command mkdir -p "$HOME/.local/share/zinit" && command chmod g-rwX "$HOME/.local/share/zinit"
>     command git clone https://github.com/zdharma-continuum/zinit "$HOME/.local/share/zinit/zinit.git" && \
>         print -P "%F{33} %F{34}Installation successful.%f%b" || \
>         print -P "%F{160} The clone has failed.%f%b"fi
> source "$HOME/.local/share/zinit/zinit.git/zinit.zsh"
> # Load a few important annexes, without Turbo
> zinit wait lucid for \
>     zdharma-continuum/zinit-annex-as-monitor \
>     zdharma-continuum/zinit-annex-bin-gem-node \
>     zdharma-continuum/zinit-annex-patch-dl \
>     zdharma-continuum/zinit-annex-rust
> ### End of Zinit's installer chunk
> # 'zdharma-continuum/fast-syntax-highlighting' Feature rich syntax highlighting for zsh
> zinit wait lucid for \
>  atinit"ZINIT[COMPINIT_OPTS]=-C; zicompinit; zicdreplay" \
>     zdharma-continuum/fast-syntax-highlighting \
>  atload"_zsh_autosuggest_start" \
>  atinit"bindkey '^_' autosuggest-execute;bindkey '^ ' autosuggest-accept;" \
>     @zsh-users/zsh-autosuggestions \
>  blockf atpull'zinit creinstall -q .' \
>     @zsh-users/zsh-completions
> # 'marlonrichert/zsh-autocomplete' - Real-time type-ahead completion for Zsh. Asynchronous find-as-you-type autocompletion.# zinit load marlonrichert/zsh-autocomplete
> # 'agkozak/zsh-z' - jump quickly to frequently and recently accessed folders
> zinit load agkozak/zsh-z
> # Refer the documentation at https://github.com/softmoth/zsh-vim-mode# 'softmoth/zsh-vim-mode' - Friendly bindings for ZSH's vi mode# zinit ice wait lucid atload'MODE_CURSOR_VIINS="bar"; vim-mode-cursor-init-hook' # zinit load softmoth/zsh-vim-mode
> # softmoth/zsh-vim-mode has been commented since zsh_init module is loaded repeatedly via a scheduler# and this is rebinding "^R" to
>
> MODE_CURSOR_VIINS="#00ff00 blinking bar"
> MODE_CURSOR_REPLACE="$MODE_CURSOR_VIINS #ff0000"
> MODE_CURSOR_VICMD="green block"
> MODE_CURSOR_SEARCH="#ff00ff steady underline"
> MODE_CURSOR_VISUAL="$MODE_CURSOR_VICMD steady bar"
> MODE_CURSOR_VLINE="$MODE_CURSOR_VISUAL #00ffff"
> # Docker
> zinit ice as"completion"
> zinit snippet https://github.com/docker/cli/blob/master/contrib/completion/zsh/_docker
>
> zinit ice from"gh-r" as"program" mv"docker* -> docker-compose"
> zinit light docker/compose
>
> This configuration does not have the snippet I suggested...? Did my
> snippet work?
>
> zinit light @zdharma-continuum/zinit-annex-bin-gem-node;
> zinit for \
>   @softmoth/zsh-vim-mode \
>        atinit'eval "$(mcfly init zsh)"' \
>        atload'mcfly_prompt_command' \
>        from'gh-r' \
>        sbin'mcfly' \
>   @cantino/mcfly;
>
> If so, please consider closing this issue and open a new one or start a
> discussion for other issues.
> ------------------------------
>
> Additionally, to restart zsh & zinit cleanly, run the command:
>
> exec zsh
>
> Try the following config:
>
> if [[ ! -f $HOME/.local/share/zinit/zinit.git/zinit.zsh ]]; then
>     print -P "%F{33} %F{220}Installing %F{33}ZDHARMA-CONTINUUM%F{220} Initiative Plugin Manager (%F{33}zdharma-continuum/zinit%F{220})…%f"
>     if command mkdir -p "$HOME/.local/share/zinit"; then
>         command chmod g-rwX "$HOME/.local/share/zinit"
>     fi
>     if command git clone https://github.com/zdharma-continuum/zinit "$HOME/.local/share/zinit/zinit.git"; then
>         print -P "%F{33} %F{34}Installation successful.%f%b"
>     else
>         print -P "%F{160} The clone has failed.%f%b"
>     fifi
> source "$HOME/.local/share/zinit/zinit.git/zinit.zsh"
> autoload _zinit && (( ${+_comps} )) && _comps[zinit]=_zinit
>
> zinit for @zdharma-continuum/zinit-annex-{'as-monitor','binary-symlink','patch-dl','rust'}
>
> MODE_CURSOR_VIINS="#00ff00 blinking bar"
> MODE_CURSOR_REPLACE="$MODE_CURSOR_VIINS #ff0000"
> MODE_CURSOR_VICMD="green block"
> MODE_CURSOR_SEARCH="#ff00ff steady underline"
> MODE_CURSOR_VISUAL="$MODE_CURSOR_VICMD steady bar"
> MODE_CURSOR_VLINE="$MODE_CURSOR_VISUAL #00ffff"
>
> zinit light-mode for \
>       atinit'zicompinit;' \
>     @marlonrichert/zsh-autocomplete \
>     @softmoth/zsh-vim-mode
>
> zinit ice as"completion"
> zinit snippet https://github.com/docker/cli/blob/master/contrib/completion/zsh/_docker
>
> zinit light-mode from"gh-r" for \
>       lbin'!*->docker-compose' \
>     @docker/compose \
>       atinit'eval "$(mcfly init zsh)"' \
>       atload'mcfly_prompt_command' \
>       lbin'!mcfly' \
>     @cantino/mcfly
>
> zinit wait'!' lucid for \
>       atinit"zicdreplay" \
>     @zdharma-continuum/fast-syntax-highlighting \
>       atload"_zsh_autosuggest_start" \
>       atinit"bindkey '^_' autosuggest-execute;bindkey '^ ' autosuggest-accept;" \
>     @zsh-users/zsh-autosuggestions \
>       atpull'zinit creinstall -q .' \
>       blockf \
>     @zsh-users/zsh-completions
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zdharma-continuum/zinit/discussions/596#discussioncomment-7870701>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIZULVJUQHICCPSF4DD2OLYJWFU5AVCNFSM6AAAAABABVU7AGVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TQNZQG4YDC>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Regards,
Ravi

**Reply by vladdoster** [reply]
@rasivasu,

There is no image attached.

**Reply by rasivasu** [reply]
Sorry, this is the error:

![image](https://github.com/zdharma-continuum/zinit/assets/1153582/96193dda-447f-420d-ae7a-9c07f5c6af8e)

**Comment by rasivasu** [comment]
@vladdoster 

I made changes incrementally this time till the shell spit out an error.  I have uploaded the file for you to look at and have pasted the error below for your reference.  I hope this is helpful.

![image](https://github.com/zdharma-continuum/zinit/assets/1153582/8fc93c75-59c2-48c8-8892-1309767578b2)


[zsh_zinit.zip](https://github.com/zdharma-continuum/zinit/files/13694798/zsh_zinit.zip)
